cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -121,28 +133,71 @@ public static void validateConfig(SecureStore secureStore, SourceControlConfig s
     }
   }
 
+  /**
+   * Returns the <a href="https://git-scm.com/docs/git-hash-object">Git Hash</a>
+   * of the requested file path in the provided commit. For symlinks, it returns the hash of
+   * the target file. This ensures that the file hash of a symlink is equal to the hash of the target file. It
+   * doesn't return the hash for directories. If the directory contains symlinks, we can't depend on the
+   * directory hash to ensure the contents are unchanged.
+   *
+   * @param relativePath The path relative to the repository base path (returned by
+   *                     {@link RepositoryManager#getRepositoryRoot()}) on the filesystem.
+   * @param commitID     The commit ID hash for which to get the file hash.
+   * @return The git file hash of the requested file.
+   * @throws IOException              when file or commit isn't found, there are circular symlinks or other file IO
+   *                                  errors.
+   * @throws IllegalArgumentException when the provided file isn't a regular file, eg: directory.
+   * @throws NotFoundException        when the file isn't committed to Git.
+   * @throws IllegalStateException    when {@link RepositoryManager#cloneRemote()} isn't called before this.
+   */
+  public String getFileHash(Path relativePath, String commitID) throws IOException, NotFoundException {
+    validateInitialized();
+    // Git considers symlinks as separate files. So the hash of the symlink will be different from the target file.
+    // Resolve path in case of symlinks. This ensures the hash of the target file is returned.
+    Path realPath = getRepositoryRoot().resolve(relativePath).toRealPath();
```

Comment:
This should be done with repository at the commitID, otherwise you may get a wrong path.

Reply:
I've made changes to checkout the commit ID and restoring the working directory at the end.